server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/SieveScriptRoutes.java

```diff
@@ -0,0 +1,173 @@
+
+    @Inject
+    public SieveScriptRoutes(SieveRepository sieveRepository, UsersRepository usersRepository) {
+        this.sieveRepository = sieveRepository;
+        this.usersRepository = usersRepository;
+    }
+
+    @Override
+    public String getBasePath() {
+        return ROOT_PATH;
+    }
+
+    @Override
+    public void define(Service service) {
+        defineAddActiveSieveScript(service);
+    }
+
+    @PUT
+    @ApiOperation(value = "Upload a new Sieve Script")
+    @Path(value = ROOT_PATH + "/{" + USER_NAME + "}/" + SCRIPT + "/{" + SCRIPT_NAME + "}")
+    @ApiResponses(value = {
+        @ApiResponse(code = HttpStatus.NO_CONTENT_204, message = "OK"),
+        @ApiResponse(code = HttpStatus.NOT_FOUND_404, message = "Invalid or non existent user"),
+        @ApiResponse(code = HttpStatus.NOT_FOUND_404, message = "Invalid Sieve script name"),
+        @ApiResponse(code = HttpStatus.NOT_FOUND_404, message = "Empty script is not accepted")
+    })
+    @ApiImplicitParams({
+        @ApiImplicitParam(
+            name = USER_NAME,
+            required = true,
+            paramType = "path",
+            dataType = "String",
+            defaultValue = "None",
+            example = "/sieve/userNameA/scripts/scriptName1",
+            value = "Username"),
+        @ApiImplicitParam(
+            name = SCRIPT_NAME,
+            required = true,
+            paramType = "path",
+            dataType = "String",
+            defaultValue = "None",
+            example = "/sieve/userNameA/scripts/scriptName1",
+            value = "Script name"),
+        @ApiImplicitParam(
+            required = false,
+            paramType = "query parameter",
+            dataType = "Boolean",
+            defaultValue = "None",
+            example = "?activate=true",
+            value = "If present, automatically activating the script.")
+    })
+    public void defineAddActiveSieveScript(Service service) {
+        service.put(USER_SCRIPT_PATH, this::addActiveSieveScript);
+    }
+
+    private HaltException addActiveSieveScript(Request request, Response response) throws UsersRepositoryException, QuotaExceededException, StorageException, ScriptNotFoundException {
+        User user = extractUser(request);
+        ScriptName script = extractScriptName(request);
+        sieveRepository.putScript(user, script, extractSieveScriptFromRequest(request));
+        if (isActivate(request.queryParams(ACTIVATE_PARAMS))) {
+            sieveRepository.setActive(user, script);
+        }
+        return halt(HttpStatus.NO_CONTENT_204);
+    }
+
+    private User extractUser(Request request) throws UsersRepositoryException {
+        return Optional.ofNullable(request.params(USER_NAME))
+            .filter(Throwing.predicate(userName -> !Strings.isNullOrEmpty(userName) && usersRepository.contains(userName)))
+            .map(User::fromUsername)
+            .orElseThrow(() -> throw404withInvalidArgument("Invalid or non existent user"));
```

Comment:
:-1: 

We should return 400 OR 404.

You need to checjthat the name is valid first (`User.fromUsername` will thow in that case) - which leads to a precise 400 error.

If the valid username do not exist, then we should generate a 404.

But we can not generate a 404 in both cases.